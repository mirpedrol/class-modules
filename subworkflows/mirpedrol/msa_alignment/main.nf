include { MAFFT } from '../../../modules/mirpedrol/mafft/main'
include { KALIGN_ALIGN } from '../../../modules/mirpedrol/kalign/align/main'
include { FAMSA_ALIGN } from '../../../modules/mirpedrol/famsa/align/main'
include { MUSCLE5_SUPER5 } from '../../../modules/mirpedrol/muscle5/super5/main'
include { MAGUS_ALIGN } from '../../../modules/mirpedrol/magus/align/main'
include { CLUSTALO_ALIGN } from '../../../modules/mirpedrol/clustalo/align/main'
include { TCOFFEE_ALIGN } from '../../../modules/mirpedrol/tcoffee/align/main'
include { LEARNMSA_ALIGN } from '../../../modules/mirpedrol/learnmsa/align/main'


workflow MSA_ALIGNMENT {

    take:
    ch_fasta

    main:
    def ch_out_alignment = Channel.empty()
    def ch_out_versions = Channel.empty()

    ch_fasta
        .branch {
            meta, fasta, tool ->
                mafft: tool == "mafft"
                    return [ meta, fasta ]
                kalign_align: tool == "kalign_align"
                    return [ meta, fasta ]
                famsa_align: tool == "famsa_align"
                    return [ meta, fasta ]
                muscle5_super5: tool == "muscle5_super5"
                    return [ meta, fasta ]
                magus_align: tool == "magus_align"
                    return [ meta, fasta ]
                clustalo_align: tool == "clustalo_align"
                    return [ meta, fasta ]
                tcoffee_align: tool == "tcoffee_align"
                    return [ meta, fasta ]
                learnmsa_align: tool == "learnmsa_align"
                    return [ meta, fasta ]
        }
        .set { ch_fasta_branch }

    MAFFT( ch_fasta_branch.mafft )
    ch_out_alignment = ch_out_alignment.mix(MAFFT.out.alignment)
    ch_out_versions = ch_out_versions.mix(MAFFT.out.versions)

    KALIGN_ALIGN( ch_fasta_branch.kalign_align )
    ch_out_alignment = ch_out_alignment.mix(KALIGN_ALIGN.out.alignment)
    ch_out_versions = ch_out_versions.mix(KALIGN_ALIGN.out.versions)

    FAMSA_ALIGN( ch_fasta_branch.famsa_align )
    ch_out_alignment = ch_out_alignment.mix(FAMSA_ALIGN.out.alignment)
    ch_out_versions = ch_out_versions.mix(FAMSA_ALIGN.out.versions)

    MUSCLE5_SUPER5( ch_fasta_branch.muscle5_super5 )
    ch_out_alignment = ch_out_alignment.mix(MUSCLE5_SUPER5.out.alignment)
    ch_out_versions = ch_out_versions.mix(MUSCLE5_SUPER5.out.versions)

    MAGUS_ALIGN( ch_fasta_branch.magus_align )
    ch_out_alignment = ch_out_alignment.mix(MAGUS_ALIGN.out.alignment)
    ch_out_versions = ch_out_versions.mix(MAGUS_ALIGN.out.versions)

    CLUSTALO_ALIGN( ch_fasta_branch.clustalo_align )
    ch_out_alignment = ch_out_alignment.mix(CLUSTALO_ALIGN.out.alignment)
    ch_out_versions = ch_out_versions.mix(CLUSTALO_ALIGN.out.versions)

    TCOFFEE_ALIGN( ch_fasta_branch.tcoffee_align )
    ch_out_alignment = ch_out_alignment.mix(TCOFFEE_ALIGN.out.alignment)
    ch_out_versions = ch_out_versions.mix(TCOFFEE_ALIGN.out.versions)

    LEARNMSA_ALIGN( ch_fasta_branch.learnmsa_align )
    ch_out_alignment = ch_out_alignment.mix(LEARNMSA_ALIGN.out.alignment)
    ch_out_versions = ch_out_versions.mix(LEARNMSA_ALIGN.out.versions)



    emit:
    alignment = ch_out_alignment
    versions = ch_out_versions

}

