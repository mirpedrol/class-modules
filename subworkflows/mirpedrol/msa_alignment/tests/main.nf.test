nextflow_workflow {

    name "Test Subworkflow MSA_ALIGNMENT"
    script "../main.nf"
    workflow "MSA_ALIGNMENT"

    tag "subworkflows"
    tag "subworkflows_mirpedrol"
    tag "subworkflows/msa_alignment"
	tag "mafft"
	tag "kalign/align"
	tag "famsa/align"
	tag "muscle5/super5"
	tag "magus/align"
	tag "clustalo/align"
	tag "tcoffee/align"
	tag "learnmsa/align"


	test("run mafft") {

		when {
			params.alignment = "mafft"
			workflow {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                        file(params.test_data['sarscov2']['illumina']['scaffolds_fasta'], checkIfExists: true)
                    ]
                """
			}
		}

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out.alignment).match("mafft_alignment")},
                { assert snapshot(workflow.out.versions).match("mafft_versions")}
            )
		}
	}

	test("run kalign/align") {

		when {
			params.alignment = "kalign/align"
			workflow {
                """
                input[0] = [ [ id:'test' ], // meta map
                          file(params.test_data['sarscov2']['illumina']['contigs_fasta'], checkIfExists: true)
                        ]
                """
			}
		}

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out.alignment).match("kalign_align_alignment")},
                { assert snapshot(workflow.out.versions).match("kalign_align_versions")}
            )
		}
	}

	test("run famsa/align") {

		when {
			params.alignment = "famsa/align"
			workflow {
                """
                input[0] = [ [ id:'test' ], // meta map
                             file(params.test_data['sarscov2']['illumina']['contigs_fasta'], checkIfExists: true)
                           ]
                """
			}
		}

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out.alignment).match("famsa_align_alignment")},
                { assert snapshot(workflow.out.versions).match("famsa_align_versions") }
            )
		}
	}

	test("run muscle5/super5") {

		when {
			params.alignment = "muscle5/super5"
			workflow {
                    """
                        input[0] = [ [ id:'test' ],
                        file("https://raw.githubusercontent.com/nf-core/test-datasets/multiplesequencealign/testdata/setoxin-ref.fa", checkIfExists: true)
                        ]
                        """
			}
		}

            then {
                assertAll(
                        { assert workflow.success },
                        { assert snapshot(workflow.out.alignment).match("muscle5_super5_alignment") },
                        { assert snapshot(workflow.out.versions).match("muscle5_super5_versions") }
                        )
		}
	}

	test("run magus/align") {

		when {
			params.alignment = "magus/align"
			workflow {
                """
                input[0] = [ [ id:'test' ], // meta map
                             file("https://raw.githubusercontent.com/nf-core/test-datasets/multiplesequencealign/testdata/setoxin-ref.fa", checkIfExists: true)
                           ]
                """
			}
		}

        then {
            assertAll(
                { assert workflow.success },
                // tests seem to be reproducible on a single machine, but not across different machines
                // test the correct samples are in there
                { assert path(workflow.out.alignment[0][1]).getTextGzip().contains(">1apf") },
                { assert path(workflow.out.alignment[0][1]).getTextGzip().contains(">1ahl") },
                { assert path(workflow.out.alignment[0][1]).getTextGzip().contains(">1atx") },
                { assert path(workflow.out.alignment[0][1]).getTextGzip().contains(">1sh1") },
                { assert path(workflow.out.alignment[0][1]).getTextGzip().contains(">1bds") },
                { assert snapshot(workflow.out.versions).match("magus_align_versions") }
            )
		}
	}

	test("run clustalo/align") {

		when {
			params.alignment = "clustalo/align"
			workflow {
                """
                input[0] = [ [ id:'test' ], // meta map
                             file(params.test_data['sarscov2']['illumina']['contigs_fasta'], checkIfExists: true)
                           ]
                """
			}
		}

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out.alignment).match("clustalo_align_alignment")},
                { assert snapshot(workflow.out.versions).match("clustalo_align_versions") }
            )
		}
	}

	test("run tcoffee/align") {

		when {
			params.alignment = "tcoffee/align"
			workflow {
                """
                input[0] = [ [ id:'test' ],
                             file("https://raw.githubusercontent.com/nf-core/test-datasets/multiplesequencealign/testdata/setoxin-ref.fa", checkIfExists: true)
                           ]
                """
			}
		}

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out.alignment).match("tcoffee_align_alignment")},
                { assert snapshot(workflow.out.versions).match("tcoffee_align_versions") }
            )
		}
	}

	test("run learnmsa/align") {

		when {
			params.alignment = "learnmsa/align"
			workflow {
                """
                input[0] = [ [ id:'test' ], // meta map
                             file(params.test_data['sarscov2']['genome']['informative_sites_fas'], checkIfExists: true)
                           ]
                """
			}
		}

        then {
            assertAll(
                { assert workflow.success },
                { assert path(workflow.out.alignment.get(0).get(1)).getTextGzip().contains(">sample1") },
                { assert snapshot(workflow.out.versions).match("learnmsa_align_versions") }
            )
		}
	}


}
