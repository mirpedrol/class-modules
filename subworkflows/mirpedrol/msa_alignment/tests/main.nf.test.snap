{
    "kalign_align_alignment": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    "test.aln.gz:md5,c165ecf48fb89862cc2a991cc3cadb2d"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.1",
            "nextflow": "24.10.0"
        },
        "timestamp": "2024-11-07T09:09:44.903862641"
    },
    "learnmsa_align_versions": {
        "content": [
            [
                "versions.yml:md5,475f61f05c9729887f723221b87de01d"
            ]
        ],
        "meta": {
            "nf-test": "0.9.1",
            "nextflow": "24.10.0"
        },
        "timestamp": "2024-11-07T09:16:10.257788333"
    },
    "clustalo_align_alignment": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    "test.aln.gz:md5,74bb9a2820a91cf68db94dbd46787722"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.1",
            "nextflow": "24.10.0"
        },
        "timestamp": "2024-11-07T09:11:39.016384316"
    },
    "kalign_align_versions": {
        "content": [
            [
                "versions.yml:md5,e7d33c95bb5d69e8573c8ad4eb2aca2c"
            ]
        ],
        "meta": {
            "nf-test": "0.9.1",
            "nextflow": "24.10.0"
        },
        "timestamp": "2024-11-07T09:09:44.918523969"
    },
    "clustalo_align_versions": {
        "content": [
            [
                "versions.yml:md5,75356bf56559adcb33a9c93aba830309"
            ]
        ],
        "meta": {
            "nf-test": "0.9.1",
            "nextflow": "24.10.0"
        },
        "timestamp": "2024-11-07T09:11:39.033642445"
    },
    "tcoffee_align_versions": {
        "content": [
            [
                "versions.yml:md5,c5208e86b43e8c973c39c2bd8ca2932a"
            ]
        ],
        "meta": {
            "nf-test": "0.9.1",
            "nextflow": "24.10.0"
        },
        "timestamp": "2024-11-07T09:12:46.142334018"
    },
    "mafft_versions": {
        "content": [
            [
                "versions.yml:md5,ac417224510279b05e804f041d82304d"
            ]
        ],
        "meta": {
            "nf-test": "0.9.1",
            "nextflow": "24.10.0"
        },
        "timestamp": "2024-11-07T09:09:24.697891518"
    },
    "tcoffee_align_alignment": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    "test.aln.gz:md5,bd1db08ad04514cc6d1334598c1a6ef0"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.1",
            "nextflow": "24.10.0"
        },
        "timestamp": "2024-11-07T09:12:46.12691554"
    },
    "famsa_align_alignment": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    "test.aln.gz:md5,7cf7375f2ba360814ea978731838b972"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.1",
            "nextflow": "24.10.0"
        },
        "timestamp": "2024-11-07T09:10:17.412909358"
    },
    "famsa_align_versions": {
        "content": [
            [
                "versions.yml:md5,c74ca8b91c442fc4ea29219ee1b724fd"
            ]
        ],
        "meta": {
            "nf-test": "0.9.1",
            "nextflow": "24.10.0"
        },
        "timestamp": "2024-11-07T09:10:17.423586217"
    },
    "muscle5_super5_alignment": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    "test.aln.gz:md5,46ba556df08f7aabbe5e1ba31d226b6a"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.1",
            "nextflow": "24.10.0"
        },
        "timestamp": "2024-11-07T09:10:32.694344"
    },
    "magus_align_versions": {
        "content": [
            [
                "versions.yml:md5,9258e7c6deb7c3d816ba75cf111e09a8"
            ]
        ],
        "meta": {
            "nf-test": "0.9.1",
            "nextflow": "24.10.0"
        },
        "timestamp": "2024-11-07T09:11:08.349010115"
    },
    "mafft_alignment": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.aln.gz:md5,23426611f4a0df532b6708f072bd445b"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.1",
            "nextflow": "24.10.0"
        },
        "timestamp": "2024-11-07T09:09:24.67999956"
    },
    "muscle5_super5_versions": {
        "content": [
            [
                "versions.yml:md5,48ec1b7cf99109e8495f3bc00d67a1eb"
            ]
        ],
        "meta": {
            "nf-test": "0.9.1",
            "nextflow": "24.10.0"
        },
        "timestamp": "2024-11-07T09:10:32.708162989"
    }
}